html5ever/macros/match_token.rs

```diff
@@ -154,82 +155,89 @@ pub struct Tag {
+        let name = if let Ok(i) = input.parse::<syn::Ident>() {
+            Some(i)
+        } else {
+            input.parse::<Token![_]>()?;
+            None
+        };
+        input.parse::<Token![>]>()?;
+        Ok(Tag {
                 kind: if closing.is_some() { TagKind::EndTag } else { TagKind::StartTag },
                 name: name
-            }
-        )
-    ));
+        })
+    }
 }
 
-impl syn::synom::Synom for LHS {
-    named!(parse -> Self, do_parse!(
-        pats: alt!(
-            syn!(syn::Pat) => { |p| LHS::Pattern(p) }
-            |
-            many0!(syn!(Tag)) => { |t| LHS::Tags(t) }
-        ) >>
-        (
-            pats
-        )
-    ));
+impl Parse for LHS {
+    fn parse(input: ParseStream) -> Result<Self> {
+        if let Ok(p) = input.fork().parse::<syn::Pat>() {
+            input.parse::<syn::Pat>().unwrap();
+            Ok(LHS::Pattern(p))
+        } else {
+            let mut tags = vec![];
+            while !input.peek(Token![=>]) {
+                tags.push(input.parse()?);
+            }
+            Ok(LHS::Tags(tags))
+        }
+    }
 }
 
-impl syn::synom::Synom for MatchTokenArm {
-    named!(parse -> Self, do_parse!(
-        binding: option!(
-            do_parse!(
-                name: syn!(syn::Ident) >>
-                punct!(@) >>
-                (
-                    name
-                )
-            )
-        ) >>
-        lhs: syn!(LHS) >>
-        punct!(=>) >>
-        rhs: do_parse!(
-                expr: alt!(
-                    expr_nosemi => { |e| RHS::Expression(e) }
-                    |
-                    syn!(syn::Expr) => { |e| RHS::Expression(e) }
-                    |
-                    keyword!(else) => { |_| RHS::Else }
-                ) >>
-                option!(punct!(,)) >>
-                (expr)
-        ) >>
-        (
-            MatchTokenArm {
-               binding,
-               lhs,
-               rhs,
-            }
-        )
-    ));
+impl Parse for MatchTokenArm {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let binding = if input.peek2(Token![@]) {
+            let binding = input.parse::<syn::Ident>()?;
+            input.parse::<Token![@]>()?;
+            Some(binding)
+        } else {
+            None
+        };
+        let lhs = input.parse::<LHS>()?;
+        input.parse::<Token![=>]>()?;
+        let rhs = if let Ok(expr) = input.fork().parse() {
+            input.parse::<syn::Expr>().unwrap();
+            RHS::Expression(expr)
+        } else if let Ok(block) = input.fork().parse::<syn::Block>() {
+            input.parse::<syn::Block>().unwrap();
+            let block = syn::ExprBlock {
+                attrs: vec![],
+                label: None,
+                block,
+            };
+            RHS::Expression(syn::Expr::Block(block))
+        } else {
+            input.parse::<Token![else]>()?;
+            RHS::Else
+        };
+        input.parse::<Option<Token![,]>>()?;
```

Comment:
This is too lenient and allows commas to be omitted where they should be required. Have the parser require a comma after `else` and after expressions that are not block, unless in the last arm.